cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -871,6 +915,49 @@ Status GetDictionaryEncodeKernel(FunctionContext* ctx,
   return Status::OK();
 }
 
+Status GetCountValuesKernel(FunctionContext* ctx,
+                            const std::shared_ptr<DataType>& type,
+                            std::unique_ptr<HashKernel>* out) {
+  std::unique_ptr<HashTable> hasher;
+
+#define COUNT_VALUES_CASE(InType)                                        \
+  case InType::type_id:                                                  \
+    hasher.reset(new CountValuesImpl<InType>(type, ctx->memory_pool())); \
+    break
+
+  switch (type->id()) {
+    COUNT_VALUES_CASE(NullType);
+    COUNT_VALUES_CASE(BooleanType);
+    COUNT_VALUES_CASE(UInt8Type);
+    COUNT_VALUES_CASE(Int8Type);
+    COUNT_VALUES_CASE(UInt16Type);
+    COUNT_VALUES_CASE(Int16Type);
+    COUNT_VALUES_CASE(UInt32Type);
+    COUNT_VALUES_CASE(Int32Type);
+    COUNT_VALUES_CASE(UInt64Type);
+    COUNT_VALUES_CASE(Int64Type);
+    COUNT_VALUES_CASE(FloatType);
+    COUNT_VALUES_CASE(DoubleType);
+    COUNT_VALUES_CASE(Date32Type);
+    COUNT_VALUES_CASE(Date64Type);
+    COUNT_VALUES_CASE(Time32Type);
+    COUNT_VALUES_CASE(Time64Type);
+    COUNT_VALUES_CASE(TimestampType);
+    COUNT_VALUES_CASE(BinaryType);
+    COUNT_VALUES_CASE(StringType);
+    COUNT_VALUES_CASE(FixedSizeBinaryType);
+    COUNT_VALUES_CASE(Decimal128Type);
```

Comment:
Is there a way to promote some code reuse with the other unary (single-argument) hash kernels?

Reply:
I think so, although I am not sure how to do it. Maybe moving everything to a macro? I am willing to try if somebody could give me some pointers on what's the best way to do it.

Reply:
The pattern described here might be useful: https://mortoray.com/2014/09/10/using-macros-to-simplify-type-visitors-and-enums/